plutus-tx/src/PlutusTx/Bool.hs

```diff
@@ -37,3 +37,28 @@ infixr 2 ||
 --
 not :: Bool -> Bool
 not a = if a then False else True
+
+{- Note [Lazy patterns on function parameters]
+Lazy patterns (~) on function parameters don't do anything if the function isn't inlined.
+This is because function applications in Plutus Tx are strict, so when passing an argument
+to a function, the argument will be evaluated immediately, regardless of whether the
+corresponding function parameter is lazy or strict. Specifically, both `f !x = body` and
+`f ~x = body` in Plutus Tx are compiled into `let f = \x -> body` in PIR.
```

Comment:
But aren't these two equivalent? And the second term becomes the first term after inlining. That's the point of this paragraph - it _should_ (although it sometimes doesn't because of GHC inlining) produce the same PIR eventually, whether function parameters are strict or lazy.

Reply:
Yes, they do become the same thing eventually, but the relevant thing is what they become _initially_, since that's what lets us observe the difference!

Okay, maybe let me phrase this differently.
```
f ~x = body
---> (GHC Core)
f x = body
---> (PIR)
f = \x -> body

f !x = body
---> (GHC Core)
f x = case x@x of { _ -> body }
---> (PIR)
f = \x -> let x = x in body
```
The two PIR bindings have the same semantics. But if GHC inlines the function at the core stage, then we get the body of the second PIR binding at the call site, which _does_ behave differently.

(I think we both agree with what's going on, I just think the explanation isn't quite clear for future people)

Reply:
Ok, I added the steps you laid out above in the Note.